src/sources/soundsourceproviderregistry.cpp

```diff
@@ -1,56 +1,193 @@
+    // Perform a linear search through the list
+    while (registrationsForFileExtension.end() != listIter) {
+        if (listIter->getProvider() == pProvider) {
+            if (listIter->getProviderPriority() != providerPriority) {
+                SoundSourceProviderRegistration registration(*listIter);
+                registrationsForFileExtension.erase(listIter);
+                insertRegistration(&registrationsForFileExtension, std::move(registration));
+            }
+            // else nothing to do
+            return; // done
+        }
+    }
+}
+
+void SoundSourceProviderRegistry::insertRegistration(
+        QList<SoundSourceProviderRegistration> *pRegistrations,
+        SoundSourceProviderRegistration registration) {
+    QList<SoundSourceProviderRegistration>::iterator listIter(
+            pRegistrations->begin());
+    // Perform a linear search through the list & insert
+    while (pRegistrations->end() != listIter) {
+        // Priority comparison with <=: New registrations will be inserted
+        // before existing registrations with equal priority, but after
+        // existing registrations with higher priority.
+        if (listIter->getProviderPriority() <= registration.getProviderPriority()) {
+            listIter = pRegistrations->insert(listIter, std::move(registration));
+            DEBUG_ASSERT(pRegistrations->end() != listIter);
+            return; // done
+        } else {
+            ++listIter; // continue loop
+        }
+    }
+    if (pRegistrations->end() == listIter) {
+        // List was empty or registration has the lowest priority
+        pRegistrations->append(std::move(registration));
+    }
 }
 
-SoundSourceProviderPointer SoundSourceProviderRegistry::registerEntry(const Entry& entry) {
-    DEBUG_ASSERT(m_supportedFileNameRegex.isEmpty());
-    DEBUG_ASSERT(entry.pProvider);
+void SoundSourceProviderRegistry::deregisterProvider(
+        const SoundSourceProviderPointer& pProvider) {
     const QStringList supportedFileExtensions(
-            entry.pProvider->getSupportedFileExtensions());
-    DEBUG_ASSERT(entry.pPluginLibrary || !supportedFileExtensions.isEmpty());
-    if (entry.pPluginLibrary && supportedFileExtensions.isEmpty()) {
-        qWarning() << "SoundSource plugin does not support any file types"
-                << entry.pPluginLibrary->getFilePath();
+            pProvider->getSupportedFileExtensions());
+    for (const auto& fileExt: supportedFileExtensions) {
+        deregisterProviderForFileExtension(fileExt, pProvider);
     }
-    foreach (const QString& supportedFileExtension, supportedFileExtensions) {
-        m_entries.insert(supportedFileExtension, entry);
+}
+
+void SoundSourceProviderRegistry::deregisterProviderForFileExtension(
+        const QString& fileExtension,
+        const SoundSourceProviderPointer& pProvider) {
+    auto mapIter(m_registry.find(fileExtension));
+    if (m_registry.end() != mapIter) {
```

Comment:
Um, no, not exactly: One loop looks for the given provider and the other one 
for the given plugin library.

On 11/25/2015 01:46 PM, Daniel Schürmann wrote:
>
> In src/sources/soundsourceproviderregistry.cpp 
> <https://github.com/mixxxdj/mixxx/pull/605#discussion_r45862401>:
>
> >      }
> > -    foreach (const QString& supportedFileExtension, supportedFileExtensions) {
> > -        m_entries.insert(supportedFileExtension, entry);
> > +}
> > +
> > +void SoundSourceProviderRegistry::deregisterProviderForFileExtension(
> > +        const QString& fileExtension,
> > +        const SoundSourceProviderPointer& pProvider) {
> > +    auto mapIter(m_registry.find(fileExtension));
> > +    if (m_registry.end() != mapIter) {
>
> This loop code is duplicated below, could we move it to a helper function?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/605/files#r45862401>.
>